### PR TITLE
doc: Add a note about backporting

### PR DESCRIPTION
**Pull request by carnhofdaki**
See laanwj's comment in #17158
https://github.com/bitcoin/bitcoin/pull/17158#issuecomment-542627090

### CODE REVIEW

**Comment by carnhofdaki** [comment]
@laanwj So what do you think about adding this note to the documentation?

**Comment by laanwj** [comment]
I think it is a good idea to add information about backporting (when to backport, how to backport) to the documentation, but I'm not sure a disconnected note only saying what *not* to do is the best way.

**Comment by carnhofdaki** [comment]
In this case we could see that the backport was clearly requested, so "when to backport" was easy. "How to backport" was as you commented an automated conflict-free cherry-pick. Which leads me to following idea of text:

-----

Backporting is done whenever the fix needs to be present in a release branch (e.g. a security issue) or on request. If the backport is trivial and can be cherry-picked automatically, just add a "Needs backport" tag to the original issue or PR. If backport needs any conflict resolution, create the PR manually and wait for comments.

The commit message of a backport should additionaly contain following two lines in the end:

```
Github-Pull: #<original issue/PR number>
Rebased-From: <sha checksum of the code already merged into master branch>
```

What do you think @laanwj ? I am not force-pushing yet and ready to close this PR if needed.

**Comment by laanwj** [comment]
Looks good to me, thanks!

Another thing to mention maybe is to only create a backport PR (when necessary) after the original PR was merged into master. This makes sure that the latest version of the code is backported, so makes review easier.

File: CONTRIBUTING.md
Comment (hebasto) [inline CONTRIBUTING.md]:
```suggestion
(e.g., a security issue) or on request.
```

File: CONTRIBUTING.md
Comment (hebasto) [inline CONTRIBUTING.md]:
```suggestion
just request a "Needs backport" label for the original PR
```

**Review by hebasto — CHANGES_REQUESTED** [review]
Concept ACK 2d169144049da82d2db81efc872cb2e2b6eddea5.

Also [backport.py](https://github.com/bitcoin-core/bitcoin-maintainer-tools#backport) script could be mentioned.

File: CONTRIBUTING.md
Comment (promag) [inline CONTRIBUTING.md]: Could mention that the PR must be for a specific branch, not master. Also, a PR may be needed for each version where the backport makes sense.

nit, drop "wait for comments"?

File: CONTRIBUTING.md
Comment (promag) [inline CONTRIBUTING.md]: `Each commit message of a backport ...`

File: CONTRIBUTING.md
Comment (promag) [inline CONTRIBUTING.md]: commit hash of the respective merged commit?

**Review by promag — COMMENTED** [review]
Concept ACK. Could link to an example PR?

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Thank you. A nice one. I actually learned something.
https://www.dailywritingtips.com/comma-after-i-e-and-e-g/

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Done in 7c5ce792ef498598bd57780d42938139dbcd58e7 and will be squashed in the end.

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Why a specific branch, not master? I understand that my development is done on a specific branch in my forked repo, but I ask for the Pull Request to master.

"wait for comments" dropped in 1d9ffdd2432f9a89f77204505b221616cf74b704 and will be squashed in the end

Comment (carnhofdaki) [reply CONTRIBUTING.md]: Backports are done when something is already in master (that is the practice of Linux development, where git itself comes from; in contrary to what Android or other BSP developers tend to do - they add specific hardware patches to a stable branch, making master run away quickly and then they are unable to catch up so after the device is not sold anymore, one can hardly find a current kernel and Android build for it...)

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Thanks. Fixed in 163fe28c54105fdde687cc49c51e635d6e483ccb and will be squashed later into the final commit.

File: CONTRIBUTING.md
Comment (hebasto) [reply CONTRIBUTING.md]: > Why a specific branch, not master? I understand that my development is done on a specific branch in my forked repo, but I ask for the Pull Request to master.

Commits are backported *from* the master branch *to* a specific branch, e.g., `0.18` or `0.19`.

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Fixed in 842aa85aa ...

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: @promag @hebasto OK, sorry. I totally misunderstood.

Comment (carnhofdaki) [reply CONTRIBUTING.md]: @promag What do you think about 2aa9503ea ?

**Comment by carnhofdaki** [comment]
> Also [backport.py](https://github.com/bitcoin-core/bitcoin-maintainer-tools#backport) script could be mentioned.

@hebasto please have a look at 28537b729 and tell me what you think

**Comment by carnhofdaki** [comment]
> Concept ACK. Could link to an example PR?

@promag I vote for backporting this PR to `0.19` and linking it in the doc as an example.
If you have other example in mind, please send a link.

**Comment by promag** [comment]
I mean adding an example, like #16189.

**Review by fanquake — COMMENTED** [review]
I'd re-write the text to be something like:

Security and bug fixes can be backported from `master` to release branches.
If the backport is non-trivial, it may be appropriate to open an additional PR,
to backport the change, only after the original PR has been merged.
Otherwise, backports will be done in batches by maintainers.

A backport should contain the following metadata in the commit body:
```
Github-Pull ...
```

Also, this PR does not need backporting.

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Done. 57379c8a5

Comment (carnhofdaki) [reply CONTRIBUTING.md]: Aah, I just realized this "Commit suggestion" button :-) Will try it next time.

**Comment by carnhofdaki** [comment]
> I mean adding an example, like #16189.

@promag Added. Have a look.

**Comment by carnhofdaki** [comment]
> I'd re-write the text to be something like:
> 
> Security and bug fixes can be backported from `master` to release branches.
> If the backport is non-trivial, it may be appropriate to open an additional PR,
> to backport the change, only after the original PR has been merged.
> Otherwise, backports will be done in batches by maintainers.
> 
> A backport should contain the following metadata in the commit body:
> 
> ```
> Github-Pull ...
> ```
> 
> Also, this PR does not need backporting.

@fanquake Text changed. Have a look.

File: CONTRIBUTING.md
Comment (luke-jr) [inline CONTRIBUTING.md]: This is intended to be a new policy?

File: CONTRIBUTING.md
Comment (luke-jr) [inline CONTRIBUTING.md]: ... unless a straightforward merge of the original branch is possible.

**Review by luke-jr — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: As far as I understood, we are just describing the commonly used, but uncodified process. Or is it something new @laanwj ?

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Why would a straightforward merge of something which has different checksum in `master` branch not contain a pointer to the parent (original commit in `master`)?

Please provide some `git` command line examples on how to find the equivalent commit in master, given a release commit. If possible and if I understand correctly. Thank you.

File: CONTRIBUTING.md
Comment (fanquake) [reply CONTRIBUTING.md]: > This is intended to be a new policy?

It's not a new policy, just what happens [in reality (when looking at bulk backports)](https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+label%3ABackport+is%3Aclosed). However if this is controversial just drop the sentence.

File: CONTRIBUTING.md
Comment (laanwj) [reply CONTRIBUTING.md]: Sometimes it's possible to base a new commit on a commit *before* the branch split (e.g. between `0.19` and `master`). If nothing conflicts in between, it can be a clean merge to both branches. This is quite rare, though, and I'm not sure mentioning it here is important.

Comment (carnhofdaki) [reply CONTRIBUTING.md]: @laanwj I see.

@luke-jr Was this ^^ what you meant?

I still think that one commit has to be the first one (in linear time) and merged into `master`. Then any other cherry-pick, no matter if there were conflicts, should IMHO contain a reference to the original unless it is easy in git to find out the original (first) patch on which this "backport" is based...

File: CONTRIBUTING.md
Comment (laanwj) [reply CONTRIBUTING.md]: Maybe mention that the `Needs backport (...)` labels are a tracking tool for maintainers and not meant as a TODO for the author of the PR.

File: CONTRIBUTING.md
Comment (laanwj) [reply CONTRIBUTING.md]: Yes, no matter how they're represented in git, changes should still be merged first into master. This is to make absolutely sure that the future major release contains everything.

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: @laanwj see 4d47775c1

Comment (carnhofdaki) [reply CONTRIBUTING.md]:
```diff
diff --git a/CONTRIBUTING.md b/CONTRIBUTING.md
index d38e44c47..f36f29dcf 100644
--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -317,7 +317,9 @@ branches.
 If the backport is non-trivial, it may be appropriate to open an
 additional PR, to backport the change, only after the original PR
 has been merged.
-Otherwise, backports will be done in batches by maintainers.
+Otherwise, backports will be done in batches by maintainers and
+the maintainers will use the proper `Needs backport (...)` labels
+when needed (the original author does not need to worry).

 A backport should contain the following metadata in the commit body:

```

File: CONTRIBUTING.md
Comment (carnhofdaki) [reply CONTRIBUTING.md]: Right. My point was just if there is a better way to find "parent" (i.e. original upstream patch) but I found there is not because even in Linux kernel development (and I suppose they know how to use git given it was born there) they always write a comment in a backported patch:

```
commit <hash> upstream.
```

See [this backported patch](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=9f8dd40c68c1) which is backport of [this original patch in master](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=3dd550a2d365) and the only difference between those two is the added comment:

```diff
--- patchorig   2019-10-30 14:00:51.058777980 +0100
+++ backport    2019-10-30 14:00:30.865990235 +0100
@@ -1,9 +1,11 @@
-commit 3dd550a2d365
+commit 9f8dd40c68c1
 Author: Alan Stern <stern@rowland.harvard.edu>
 Date:   Wed Sep 4 11:56:27 2019 -0400

     USB: usbcore: Fix slab-out-of-bounds bug during device reset

+    commit 3dd550a2d36596a1b0ee7955da3b611c031d3873 upstream.
+
     The syzbot fuzzer provoked a slab-out-of-bounds error in the USB core:

     BUG: KASAN: slab-out-of-bounds in memcmp+0xa6/0xb0 lib/string.c:904
```

So maybe I totally misunderstood @luke-jr and was solving some non-issue here, but I learned a lot. Thanks!

Comment (luke-jr) [reply CONTRIBUTING.md]: Yes, bugfixes are best when based immediately on top of the commit introducing the bug, in which case it is a clean merge to all affected branches.

Sometimes that's not possible and a cherry-pick is needed instead, but when it is possible, it should be preferred.

(Linux doesn't set a good example in this regard.)

File: CONTRIBUTING.md
Comment (luke-jr) [inline CONTRIBUTING.md]: Maybe just remove "maintainers" here?

Comment (carnhofdaki) [reply CONTRIBUTING.md]: https://github.com/carnhofdaki/bitcoin/commit/9e35eaeddbb46405e30c05de27ebdb7aa4e87fd1

**Comment by fanquake** [comment]
This looks ok now. @carnhofdaki  can you squash.

**Comment by carnhofdaki** [comment]
@fanquake Sure. Thank you for info.

**Pull request merged**
